functions/federated_aws_rp/utils.py

```diff
@@ -0,0 +1,298 @@
+    :return: A base64 encoded string with padding stripped
+    """
+    return base64.urlsafe_b64encode(data).decode("utf-8").rstrip("=")
+
+
+def encode_cookie_value(value: dict) -> str:
+    """Given a dictionary, JSON dump it and encode the result for a cookie
+
+    :param value: A dictionary to encode
+    :return: An encoded string to be set in a cookie
+    """
+    return http.cookies.SimpleCookie().value_encode(
+        json.dumps(value, separators=(',', ':')))[1]
+
+
+def get_store(cookie_header: str) -> dict:
+    """Given an HTTP cookie header return the store
+
+    Extract the cookie with the name CONFIG.cookie_name from the cookie_header
+    parse the JSON from it and return it
+
+    :param cookie_header: The content of the "Cookie" HTTP header
+    :return: The value of the cookie
+    """
+    cookies = http.cookies.SimpleCookie()
+    cookies.load(cookie_header)
+    if CONFIG.cookie_name in cookies:
+        try:
+            return json.loads(cookies[CONFIG.cookie_name].value)
+        except Exception:
+            logger.error('Unable to parse  JSON from "{}"'.format(
+                cookies[CONFIG.cookie_name].value
+            ))
+            return {}
+    else:
+        return {}
+
+
+def get_discovery_document(discovery_url: str) -> dict:
+    """Fetch the OIDC discovery document and enrich it with the JWKS
+
+    Fetch the the discovery_url and parse it into a dictionary. Fetch the
+    content of the jwks_uri and add it to the dictionary in the "jkws" key
+
+    :param discovery_url: The URL of the identity providers OIDC discovery url
+    :return: A dictionary containing information about the identity provider
+    """
+    result = requests.get(discovery_url).json()
+    result['jwks'] = requests.get(result['jwks_uri']).json()
+    return result
+
+
+def exchange_code_for_token(
+        code_verifier: str,
+        token_endpoint: str,
+        code: str) -> dict:
+    """Request an OIDC token from the token endpoint using the code provided
+
+    Submit the code and code_verifier to the token endpoint and return a token
+    dictionary which contains the ID token
+
+    :param dict code_verifier: The original code_verifier
+    :param str token_endpoint: The URL of the token endpoint
+    :param str code: A string returned from the identity provider
+    :return: A dictionary containing the encoded ID token
+    """
+    headers = {
+        "Content-Type": "application/json"
+    }
+    body = {
+        "grant_type": "authorization_code",
+        "client_id": CONFIG.client_id,
+        "code_verifier": code_verifier,
+        "code": code,
+        "redirect_uri": CONFIG.redirect_uri,
+    }
+
+    # TODO : Wait should we not do PKCE because this is a traditional RP
+    # and instead use a client secret?
+    logger.debug('Requesting token with arguments {}'.format(body))
+    token = requests.post(
+        token_endpoint,
+        headers=headers,
+        json=body).json()
+    return token
```

Comment:
I would consolidate these previous two statements.

Reply:
Fixed in f98e10db3beaf7d98f4002c0da6a6ec6684814b2